rpk shadow: add full configuration template with field documentation

r-vasquez opened the pull request:
This PR introduces the new flag `--print-template` to `rpk shadow config generate` that prints a full configuration template, including field documentation, so our users can understand every field properly.

To achieve this, we are adding a new tool: go-pbgen, which is a protoc plugin that extracts proto field comments and generates a `comments.pb.go` file for runtime access from rpk. This also adds a new CI job to catch drifts.

Example output:

`rpk shadow config generate --print-template`
```yaml
# Shadow Link Configuration Template
# Configuration for the internal kafka client
client_options:
  # The bootstrap servers to use
  bootstrap_servers: []
  # If provided, this is the expected ID of the source cluster.  If it does
  # not match then the connection will be rejected.  If provided, this value
  # must match the `ClusterId` field returned in the Kafka Metadata response
  # message
  source_cluster_id: ""
  # TLS settings
  tls_settings:
    # Whether or not TLS is enabled
    enabled: false
    # Certificates and keys are provided as files
    tls_file_settings:
      # Path to the CA
      ca_path: ""
      # Key and Cert are optional but if one is provided, then both must be
      # Path to the key
      key_path: ""
      # Path to the cert
      cert_path: ""
    # Certificates and keys are provided in PEM format
    tls_pem_settings:
      # The CA
      ca: ""
      # Key and Cert are optional but if one is provided, then both must be
      # The key
      key: ""
      # The cert
      cert: ""
    # If true, the SNI hostname will not be provided when TLS is used
    do_not_set_sni_hostname: false
  # Authentication settings
  authentication_configuration:
    # SASL/SCRAM configuration
    scram_configuration:
      # SCRAM username
      username: ""
      # Password
      password: ""
      # The SCRAM mechanism to use
      scram_mechanism: SCRAM_SHA_256 # SCRAM-SHA-256
  # Max metadata age.
  # If 0 is provided, defaults to 10 seconds
  metadata_max_age_ms: 0
  # Connection timeout.
  # If 0 is provided, defaults to 1 second
  connection_timeout_ms: 0
  # Retry base backoff.
  # If 0 is provided, defaults to 100ms
  retry_backoff_ms: 0
  # Fetch request timeout.
  # If 0 is provided, defaults to 500ms
  fetch_wait_max_ms: 0
  # Fetch min bytes.
  # If 0 is provided, defaults to 5 MiB
  fetch_min_bytes: 0
  # Fetch max bytes.
  # If 0 is provided, defaults to 20 MiB
  fetch_max_bytes: 0
  # Fetch partition max bytes.
  # If 0 is provided, defaults to 1 MiB
  fetch_partition_max_bytes: 0
# Topic metadata sync options
topic_metadata_sync_options:
  # How often to sync metadata
  # If 0 provided, defaults to 30 seconds
  interval: 30s # duration (e.g., 30s, 1m, 1h)
  # List of filters that indicate which topics should be automatically
  # created as shadow topics on the shadow cluster.  This only controls
  # automatic creation of shadow topics and does not effect the state of the
  # mirror topic once it is created.
  # Literal filters for __consumer_offsets, _redpanda.audit_log and _schemas
  # will be rejected as well as prefix filters to match topics prefixed with
  # _redpanda or __redpanda.
  # Wildcard `*` is permitted only for literal filters and will _not_ match
  # any topics that start with _redpanda or __redpanda.  If users wish to
  # shadow topics that start with _redpanda or __redpanda, they should
  # provide a literal filter for those topics.
  auto_create_shadow_topic_filters:
    # Literal or prefix
    pattern_type: LITERAL # Must match the filter exactly
    # Include or exclude
    filter_type: INCLUDE # Include the items that match the filter
    # The resource name, or "*"
    # Note if "*", must be the _only_ character
    # and `pattern_type` must be `PATTERN_TYPE_LITERAL`
    name: ""
  # List of topic properties that should be synced from the source topic.
  # The following properties will always be replicated
  # - Partition count
  # - `max.message.bytes`
  # - `cleanup.policy`
  # - `timestamp.type`
  # The following properties are not allowed to be replicated and adding them
  # to this list will result in an error:
  # - `redpanda.remote.readreplica`
  # - `redpanda.remote.recovery`
  # - `redpanda.remote.allowgaps`
  # - `redpanda.virtual.cluster.id`
  # - `redpanda.leaders.preference`
  # - `redpanda.cloud_topic.enabled`
  # This list is a list of properties in addition to the default properties
  # that will be synced.  See `exclude_default`.
  synced_shadow_topic_properties: []
  # If false, then the following topic properties will be synced by default:
  # - `compression.type`
  # - `retention.bytes`
  # - `retention.ms`
  # - `delete.retention.ms`
  # - Replication Factor
  # - `min.compaction.lag.ms`
  # - `max.compaction.lag.ms`
  # If this is true, then only the properties listed in
  # `synced_shadow_topic_properties` will be synced.
  exclude_default: false
  # Enables data replication from the earliest offset
  # on the source topic/partition.
  start_at_earliest:
  # Enables data replication from the latest offset
  # on the source topic/partition.
  start_at_latest:
  # Enables data replication from the first offset on the
  # source topic/partition where the record's timestamp is
  # at or after the specified timestamp.
  start_at_timestamp: "2024-01-01T00:00:00Z" # RFC3339 timestamp
  # Allows user to pause the topic sync task.  If paused, then
  # the task will enter the 'paused' state and not sync topics or their
  # properties from the source cluster
  paused: false
# Consumer offset sync options
consumer_offset_sync_options:
  # Sync interval
  # If 0 provided, defaults to 30 seconds
  interval: 30s # duration (e.g., 30s, 1m, 1h)
  # Allows user to pause the consumer offset sync task.  If paused, then
  # the task will enter the 'paused' state and not sync consumer offsets from
  # the source cluster
  paused: false
  # The filters
  group_filters:
    # Literal or prefix
    pattern_type: LITERAL # Must match the filter exactly
    # Include or exclude
    filter_type: INCLUDE # Include the items that match the filter
    # The resource name, or "*"
    # Note if "*", must be the _only_ character
    # and `pattern_type` must be `PATTERN_TYPE_LITERAL`
    name: ""
# Security settings sync options
security_sync_options:
  # Sync interval
  # If 0 provided, defaults to 30 seconds
  interval: 30s # duration (e.g., 30s, 1m, 1h)
  # Allows user to pause the security settings sync task.  If paused,
  # then the task will enter the 'paused' state and will not sync security
  # settings from the source cluster
  paused: false
  # ACL filters
  acl_filters:
    # The resource filter
    resource_filter:
      # The ACL resource type to match
      resource_type: ANY
      # The pattern to apply to name
      pattern_type: ANY
      # Name, if not given will default to match all items in `resource_type`.
      # Note that asterisk `*` is literal and matches resource ACLs
      # that are named `*`
      name: ""
    # The access filter
    access_filter:
      # The name of the principal, if not set will default to match
      # all principals with the specified `operation` and `permission_type`
      principal: ""
      # The ACL operation to match
      operation: ANY
      # The permission type
      permission_type: ANY
      # The host to match.  If not set, will default to match all hosts
      # with the specified `operation` and `permission_type`. Note that
      # the asterisk `*` is literal and matches hosts that are set to `*`
      host: ""
# Schema Registry sync options
schema_registry_sync_options:
  shadow_schema_registry_topic:
 ```

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [X] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.3.x
- [ ] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x

## Release Notes

* none

rockwotj commented on src/go/rpk/gen/protocomments/common/v1/comments.pb.go:
```diff
@@ -0,0 +1,73 @@
+	"redpanda.core.common.v1.ACLOperation":        "/ The ACL operation to match",
+	"redpanda.core.common.v1.ACLPattern":          "/ The ACL pattern type",
+	"redpanda.core.common.v1.ACLPermissionType":   "/ ACL permission types",
+	"redpanda.core.common.v1.ACLResource":         "/ The ACL resource types",
```
Can you add a commit to remove the triple comment

r-vasquez replied:
good catch, I will use this to test the CI changes.

Done, see https://github.com/redpanda-data/redpanda/pull/28350#issuecomment-3492994578

rockwotj commented on buf.gen.yaml:
```diff
@@ -0,0 +1,17 @@
+    out: src/go/rpk/gen/protocomments
+    opt:
+      - paths=import
+      - Mproto/redpanda/core/admin/v2/shadow_link.proto=github.com/redpanda-data/redpanda/src/go/rpk/gen/protocomments/admin/v2
+      - Mproto/redpanda/core/admin/v2/broker.proto=github.com/redpanda-data/redpanda/src/go/rpk/gen/protocomments/admin/v2
+      - Mproto/redpanda/core/admin/v2/cluster.proto=github.com/redpanda-data/redpanda/src/go/rpk/gen/protocomments/admin/v2
+      - Mproto/redpanda/core/admin/v2/kafka_connections.proto=github.com/redpanda-data/redpanda/src/go/rpk/gen/protocomments/admin/v2
+      - Mproto/redpanda/core/common/v1/acl.proto=github.com/redpanda-data/redpanda/src/go/rpk/gen/protocomments/common/v1
+      - Mproto/redpanda/core/common/v1/ntp.proto=github.com/redpanda-data/redpanda/src/go/rpk/gen/protocomments/common/v1
+      - Mproto/redpanda/core/common/v1/tls.proto=github.com/redpanda-data/redpanda/src/go/rpk/gen/protocomments/common/v1
```
should we just encode this into the plugin this translation?

r-vasquez commented:
Added: https://github.com/redpanda-data/redpanda/pull/28350/commits/f18d310049f10300addd04b543c74008940d3da5. This should trigger a CI error. Testing.

Sample of the error, working as expected: https://github.com/redpanda-data/redpanda/actions/runs/19114059599/job/54618329554?pr=28350

r-vasquez committed:
common/acl/proto: Remove triple comment /

r-vasquez committed:
rpk: add proto comment extraction plugin

Adds go-pbgen, a custom protoc plugin that
extracts proto field comments and generates
comments.pb.go files for runtime access.

Can be used by 'rpk shadow config generate'
to display field documentation in config
templates.

This commit also includes a new CI check
to detect proto changes.

r-vasquez committed:
rpk shadow config: generate full template

With field documentation. This uses the
newly proto-generated comment maps to
go over each field in the Shadow Link
proto and print a yaml with each
field with the proto comment on top of it.

r-vasquez committed:
rpk shadow template: remove proto prefixes

rpk don't accept the raw prefixed enums, printing
a template with them will be misleading.

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--37e46410-25eb-4180-b03f-f69d207fd37e-->
<!--TEST RESULTS-->
<!--cfa5ba8d-cb0e-4682-96b1-a09e8d5b0a5a-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/75704">build#75704</a></summary>


| test_class                          | test_method                     | test_arguments                                                                    | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  | test_history                                                                                                                                                           |
|:------------------------------------|:--------------------------------|:----------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ShadowLinkingReplicationTests       | test_replication_basic          | {"shuffle_leadership": true, "source_cluster_spec": {"cluster_type": "redpanda"}} | integration | https://buildkite.com/redpanda/redpanda/builds/75704#019a564d-b911-4404-80aa-ab3bc7e492db | FLAKY         | 16/21    | upstream reliability is '97.84263959390863'. current run reliability is '76.19047619047619'. drift is 21.65216 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingReplicationTests&test_method=test_replication_basic     |
| PartitionBalancerTest               | test_fuzz_admin_ops             | null                                                                              | integration | https://buildkite.com/redpanda/redpanda/builds/75704#019a5645-a023-4a03-86ab-6933bbbc85b5 | FLAKY         | 19/21    | upstream reliability is '100.0'. current run reliability is '90.47619047619048'. drift is 9.52381 and the allowed drift is set to 50. The test should PASS              | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=PartitionBalancerTest&test_method=test_fuzz_admin_ops                |
| RedpandaNodeOperationsSmokeTest     | test_node_ops_smoke_test        | {"cloud_storage_type": 1, "mixed_versions": false}                                | integration | https://buildkite.com/redpanda/redpanda/builds/75704#019a564d-b918-4231-a418-11fefd198c11 | FLAKY         | 11/21    | upstream reliability is '97.26443768996961'. current run reliability is '52.38095238095239'. drift is 44.88349 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=RedpandaNodeOperationsSmokeTest&test_method=test_node_ops_smoke_test |
| ScalingUpTest                       | test_moves_with_local_retention | {"use_topic_property": true}                                                      | integration | https://buildkite.com/redpanda/redpanda/builds/75704#019a5645-a025-4b67-9792-5fa456e57c2d | FLAKY         | 19/21    | upstream reliability is '97.6878612716763'. current run reliability is '90.47619047619048'. drift is 7.21167 and the allowed drift is set to 50. The test should PASS   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ScalingUpTest&test_method=test_moves_with_local_retention            |
| WriteCachingFailureInjectionE2ETest | test_crash_all                  | {"use_transactions": false}                                                       | integration | https://buildkite.com/redpanda/redpanda/builds/75704#019a5645-a024-447d-a13e-e0c96210a70a | FLAKY         | 16/21    | upstream reliability is '89.30434782608697'. current run reliability is '76.19047619047619'. drift is 13.11387 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=WriteCachingFailureInjectionE2ETest&test_method=test_crash_all       |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/75793">build#75793</a></summary>


| test_class                    | test_method                                  | test_arguments                                                                    | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  | test_history                                                                                                                                                               |
|:------------------------------|:---------------------------------------------|:----------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:---------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ShadowLinkingReplicationTests | test_replication_basic                       | {"shuffle_leadership": true, "source_cluster_spec": {"cluster_type": "redpanda"}} | integration | https://buildkite.com/redpanda/redpanda/builds/75793#019a5ba9-09e8-44f4-8969-0e0ded8dd0cf | FLAKY         | 17/21    | upstream reliability is '96.16279069767441'. current run reliability is '80.95238095238095'. drift is 15.21041 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingReplicationTests&test_method=test_replication_basic         |
| ShadowLinkingReplicationTests | test_replication_with_failures               | null                                                                              | integration | https://buildkite.com/redpanda/redpanda/builds/75793#019a5ba5-d62f-469f-95e4-a3244212cb06 | FLAKY         | 20/21    | upstream reliability is '98.40182648401826'. current run reliability is '95.23809523809523'. drift is 3.16373 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingReplicationTests&test_method=test_replication_with_failures |
| DataMigrationsApiTest         | test_creating_and_listing_migrations         | null                                                                              | integration | https://buildkite.com/redpanda/redpanda/builds/75793#019a5ba5-d62d-4c7f-a67c-e3727eb81844 | FLAKY         | 18/21    | upstream reliability is '94.34464404524284'. current run reliability is '85.71428571428571'. drift is 8.63036 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DataMigrationsApiTest&test_method=test_creating_and_listing_migrations   |
| SegmentMsTest                 | test_segment_rolling_with_retention_consumer | null                                                                              | integration | https://buildkite.com/redpanda/redpanda/builds/75793#019a5ba5-d62e-45bd-aa2f-949a7cdf67b9 | FLAKY         | 19/21    | upstream reliability is '94.26751592356688'. current run reliability is '90.47619047619048'. drift is 3.79133 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=SegmentMsTest&test_method=test_segment_rolling_with_retention_consumer   |
</details>
<!--/body-->

rockwotj commented on buf.gen.yaml:
```diff
@@ -0,0 +1,17 @@
+plugins:
+  # Custom comment extraction plugin - ONLY generate comments, not .pb.go files
+  # All .pb.go files come from buf.build/gen/go/redpandadata/core/protocolbuffers/go
+  - local: tools/go-pbgen/run-plugin.sh
```
The protos are being generated from the local proto files or from the ones in buf?

r-vasquez replied:
The comments one come from proto, the rest of the types/methods we use in rpk come from buf, maybe I should rewrite this to reflect it better

rockwotj replied:
Well can we tell the local plugin to generate comments from the same protos that RPK is using?

I'd like to avoid core needing to regenerate protos for RPK when they change them (core largely automates this with Bazel). So if RPK only has to update the protos when the buf import is updated that would be ideal from my prospective...

We can punt this to a followup if needed

r-vasquez replied:
That would be great, I'll investigate how to do it and possibly send something today. Thanks!

rockwotj commented on tools/go-comment-pbgen/README.md:
Can we call this directory `go-comment-pbgen`?

rockwotj commented on src/go/rpk/pkg/cli/shadow/config.go:
```diff
@@ -162,3 +190,259 @@ func generateSampleConfig() *ShadowLinkConfig {
+	var msgDesc protoreflect.MessageDescriptor
+	msgType, err := protoregistry.GlobalTypes.FindMessageByName("redpanda.core.admin.v2.ShadowLinkConfigurations")
+	if err != nil {
+		// Fallback: try using ProtoReflect
```
When does this fallback get hit? Shouldn't we always do it this way, because this is how we parse the protobuf?

r-vasquez replied:
You are right, this Fallback is not needed. I will just use 
```
	cfg := &v2.ShadowLinkConfigurations{}
	msg := cfg.ProtoReflect()
	if msg == nil {
		return fmt.Sprintf("# Error: Could not find message descriptor")
	}
	msgDesc := msg.Descriptor()
```

r-vasquez committed:
go-pbgen: rename to go-comment-pbgen
